Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -1205,9 +1214,39 @@ extension MacroApplication {
 // MARK: Freestanding macro expansion
 
 extension MacroApplication {
+  class MacroExpansion<ResultType> {
+    private let expandedNode: ResultType
+    private let macro: any Macro.Type
+    private unowned let macroApplication: MacroApplication
+    private var isConsumed = false
+
+    fileprivate init(expandedNode: ResultType, macro: any Macro.Type, macroApplication: MacroApplication) {
+      self.expandedNode = expandedNode
+      self.macro = macro
+      self.macroApplication = macroApplication
+    }
+
+    /// Invokes the given closure with the node resulting from a macro expansion.
+    ///
+    /// - Precondition: This method has never been called for this instance.
+    ///
+    /// This method inserts a pair of push and pop operations immediately around the closure invocation to maintain an
+    /// exact stack of expanding freestanding macros to detect circluar macro expansion. Callers should only call this
+    /// method once and perform all further expansions on `expanded` only within the scope of `body`.
+    func consume<T>(_ body: (_ expanded: ResultType) throws -> T) rethrows -> T {
+      precondition(!isConsumed)
+      isConsumed = true
```

Comment:
Sorry, should have noticed this earlier but if we have the `append` and `removeLast` paired up like this, I don’t it’s not as fatal anymore if a `MacroExpansion` is consumed twice – the state would still be left intact. So, I guess we don’t need the `isConsumed` check anymore, which removes one potential cause of crashes – the parser tries to be very fault-tolerant and never crash. And then, this could also become a `struct` again.